man/qscand.1
.\" qscand is released under the GNU GENERAL PUBLIC LICENSE.
.TH "qscand" "1" "30. December 2009" "Gennady ShultZ Kozlov" "qscand 0.1"
.SH "NAME"
qscand \- network wrapper for qscan
.SH "SYNOPSIS"
\fBqscand [-h]
.br
\fBqscand [-n] [-d]
.br
\fBqscand -i NW-INTERFACE-ADDRESS -p LISTEN-PORT \fR
.br
.SH "DESCRIPTION"
qscand is a network wrapper which gives you access to your devices through the TCP/IP stacK.

.SH "OPTIONS"

\fB\-h, --help\fR
.br
		show help

\fB\-n, --nodaemon\fR
.br
		don't daemonize qscand

\fB\-d, --debug\fR
.br
		run qscand in debug mode

\fB\-i, --iface\fR
.br
		select ip address for listening

\fB\-p, --port\fR
.br
		select port for listening

\fR \fB
\fR \fB
\fR \fB
\fRplease report man page improvements to T.Maguin@web.de\fB


debian/changelog
qpxtool (0.7.2-4.1-ArtKarMOD) unstable; urgency=medium

  * support for BD-RE WH16NS58
  * other changes

 -- ArtKar <artkar@intercore.com.pl>  Mon, 15 Apr 2019 21:19:15 +0200

qpxtool (0.7.2-4.1) unstable; urgency=medium

  * Non-maintainer upload.
  * Add patch 07-fix-wr-mode-declarations.patch:
    - Fix declaration of char attributes in wr_mode. (Closes: #856231)

 -- John Paul Adrian Glaubitz <glaubitz@physik.fu-berlin.de>  Mon, 27 Feb 2017 14:04:55 +0100

qpxtool (0.7.2-4) unstable; urgency=medium

  * Update debian/rules: delete option --parallel from dh command because it
    causes FTBFS on buildd hosts where parallel build is realy enabled.

 -- Boris Pek <tehnick@debian.org>  Sun, 13 Apr 2014 00:18:54 +0400

qpxtool (0.7.2-3) unstable; urgency=medium

  * Simplify patch 05-add-hardening-flags-in-compiler-options.patch.
  * Update debian/rules:
    - simplify rules after changes in related patch
    - add options --parallel and --list-missing to dh
  * Update debian/qpxtool.install.

 -- Boris Pek <tehnick@debian.org>  Sat, 12 Apr 2014 20:25:56 +0400

qpxtool (0.7.2-2) unstable; urgency=medium

  * Update debian/control:
    - udate using wrap-and-sort tool
    - delete now useless sections "Breaks: qpxtool (<< 0.7.1.002-1)"
    - add build dependency from hurd-dev on hurd-i386
  * Add patch 06-make-port-for-hurd-i386.patch:
    Make port for Debian GNU/Hurd.
  * Update debian/copyright.

 -- Boris Pek <tehnick@debian.org>  Sat, 12 Apr 2014 16:19:23 +0400

qpxtool (0.7.2-1) unstable; urgency=medium

  * New maintainer. (Closes: #629599)
  * Update to stable release 0.7.2.
  * Bump Standards-Version to 3.9.5 (was 3.9.3).
  * Update patches
    - add 01-do_not_use_lib64_dir.patch
    - rewrite 03-ftbfs_kfreebsd.patch
    - rewrite 02-desktop_file.patch
    - update 04-fix-build-with-libpng15.patch
    - delete 01-makefile.patch (it is unnecessary now)
  * Update debian/rules:
    - add necessary options to configure script for installing files
      in according with Debian Policy
    - add get-orig-source section
    - simplify rules
  * Fix lintian notes vcs-field-not-canonical.
  * Update years in debian/copyright.
  * Update debian/qpxtool.docs.
  * Add debian/qpxtool.doc-base.
  * Fix a lot of lintian notes hyphen-used-as-minus-sign:
    - add fixed debian/man/* files
    - add debian/qpxtool.manpages and debian/pxfw.manpages
    - update debian/qpxtool.install and debian/pxfw.install
  * Update debian/libqpx0.lintian-overrides.

 -- Boris Pek <tehnick@debian.org>  Wed, 26 Feb 2014 16:26:22 +0400

qpxtool (0.7.1.002-6) unstable; urgency=low

  * QA upload.
  * Bumped Standards-Version to 3.9.3 (was 3.9.2).
  * Bumped debhelper version to 9 (was 7.0.50~); updated debian/compat.
  * Added file debian/patches/05-add-hardening-flags-in-compiler-options.patch.
  * Updated debian/rules: added hardening flags in compiler options.
  * Changed build dependency from libpng12-dev to libpng-dev.
    Added file debian/patches/04-fix-build-with-libpng15.patch.
    Added small hack in debian/rules (see LIBPNG_VER variable).
    Build with libpng version 1.5.10 was tested successfully.
    (Closes: #662481, #648127)
  * Used [kfreebsd-any] instead of hardcoded list of kFreeBSD architectures
    [kfreebsd-i386 kfreebsd-amd64] in build dependency. (Closes: #634714)
  * File debian/copyright was updated in according to Copyright format 1.0.
  * Fixed such lintian warnings and notes:
    - out-of-date-copyright-format-uri
    - hardening-no-relro and hardening-no-fortify-functions

 -- Boris Pek <tehnick-8@mail.ru>  Tue, 26 Jun 2012 03:11:10 +0300

qpxtool (0.7.1.002-5) unstable; urgency=low

  * QA upload.
  * Bump Standards.
  * Orphaning this.

 -- Alessio Treglia <alessio@debian.org>  Tue, 21 Jun 2011 13:04:12 +0200

qpxtool (0.7.1.002-4) unstable; urgency=low

  * We need both Breaks and Replaces fields.

 -- Alessio Treglia <alessio@debian.org>  Wed, 06 Oct 2010 12:48:26 +0200

qpxtool (0.7.1.002-3) unstable; urgency=low

  * Link against libcam on kfreebsd to fix FTBFS.
  * debian/control:
    - Properly replace Conflicts/Replaces with Breaks fields.
    - Bump Standards.
  * Update debian/copyright.

 -- Alessio Treglia <alessio@debian.org>  Wed, 06 Oct 2010 09:09:15 +0200

qpxtool (0.7.1.002-2) unstable; urgency=low

  * Correct little typo in the descriptions, thanks to Davide Prina for
    reporting this (Closes: #582445).
  * Change my email address.

 -- Alessio Treglia <alessio@debian.org>  Sun, 06 Jun 2010 19:48:46 +0200

qpxtool (0.7.1.002-1) unstable; urgency=low

  * New maintainer (Closes: #543879).
  * New upstream release.
  * Move packaging to collab-maint's git area.
  * Split libraries:
    - libqpx0 (shared libs)
    - libqpx-dev (development files)
  * debian/control:
    - Drop libqt3-mt-dev, it's no longer needed.
    - Build-depend on qmake-qt4.
    - Bump Standards.
    - Set DM-Upload-Allowed to yes.
    - Build-depend on libqt4-dev.
    - Build-depend on libpng12-dev.
    - Set Conflicts/Replaces proper fields.
  * Switch to 3.0 (quilt) format.
  * Switch to debhelper 7.
  * Drop dpatch support:
    - Drop 02-typos.dpatch patch, no longer needed.
    - Drop 03-0.6.1.1.dpatch, no longer needed.
    - Drop 04-libata.dpatch patch, no longer needed.
    - 01-qtdir.dpatch -> 01-makefile.patch:
      + Drop QTDIR which seems to be not respected.
      + Set QMAKE4 properly.
      + Add DEP-3 compliant tags.
  * debian/patches/02-desktop_file.patch:
    - Fix desktop file as per spec.
  * debian/{libqpx0,qpxtool}.lintian-overrides:
    - "MultiMediaMasters & Machinary" is a manufacturer's name and not a
      spelling error.
  * debian/rules:
    - Cleanup.
    - Fix clean rule by passing the QMAKE4 envar.
  * Add watch file.

 -- Alessio Treglia <quadrispro@ubuntu.com>  Mon, 26 Apr 2010 09:36:21 +0200

qpxtool (0.6.1-5) unstable; urgency=low

  * Updating package to standards version 3.8.3.
  * Removing vcs fields.
  * Orphaning package.


console/qscand/qscand.cpp
/*
 * This file is part of the QPxTool project.
 * Copyright (C) 2008-2009 Gennady "ShultZ" Kozlov <qpxtool@mail.ru>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * See the file "COPYING" for the exact licensing terms.
 *
 */

#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "qscand.h"
#include <child.h>
#include <common_functions.h>
#include <threads.h>
#ifndef _WIN32
#include <signal.h>
#endif

#include <fcntl.h>

bool term;
int childl;
pid_t pid;
// pid_t cpids[CLIENTS_MAX];
bool daemonized = 0;
bool debug = 0;

#if defined(__unix) || defined(__unix__)
void sigint_handler(int)
#elif defined(_WIN32)
BOOL WINAPI sigint_handler(DWORD)
#endif
{
  if (!childl) {
    if (debug && !daemonized && clients > 0)
      printf("%d: SIGINT. %d client(s) left. Terminating connections...\n", pid,
             clients);
    //		while (clients) msleep((1 << 8));
    //		if (debug && !daemonized)
    //			printf("%d: SIGINT. Terminating listener...\n", pid);
  } else if (childl == 1) {
    if (debug && !daemonized)
      printf("%d: SIGINT. Terminating client connection...\n", pid);
  }
  term = 1;
#if defined(_WIN32)
  return 1;
#endif
};

#ifdef DAEMON_EN
int daemonize() {
  pid_t pid;
  switch (pid = fork()) {
  case -1:
    return -1;
  case 0:
    openlog("qscand", LOG_PID, LOG_DAEMON);
    close(0);
    close(1);
    close(2);
    daemonized = 1;
    break;
  default:
    exit(0);
  }
  if (setsid() == -1)
    return -1;
  return 0;
}
#endif

static struct option long_options[] = {{"help", 0, NULL, 'h'},
#ifdef DAEMON_EN
                                       {"nodaemon", 0, NULL, 'n'},
#endif
                                       {"debug", 0, NULL, 'd'},
                                       {"iface", 1, NULL, 'i'},
                                       {"port", 1, NULL, 'p'},
#ifdef AUTOPORT_ENABLE
                                       {"portauto", 0, NULL, 'a'},
#endif
                                       {0, 0, 0, 0}};

int child_create(int idx) {
  int r;
  if ((r = thread_create(&childs[idx].tid, NULL, &child_thread,
                         (void *)&childs[idx].arg))) {
#ifdef DAEMON_EN
    syslog(LOG_ERR, "Error creating thread!\n");
#else
    printf("Error creating thread!\n");
#endif
    return r;
    close(childs[idx].arg.connfd);
  }
  return 0;
}

int main(int argc, char **argv) {
  //	pid_t cpid;
  char *iface = NULL;
  uint32_t ip;
  uint16_t port;
  int listenfd;
  struct sockaddr_in srvaddr;
  socklen_t srvaddr_len = sizeof(srvaddr);
  int cli_idx;
#ifdef _WIN32
  WSADATA WSAdata;
#endif

  bool portauto = 0;
  bool nodaemon = 0;
  daemonized = 0;

  struct timeval tv;
  fd_set rd_set;
  FD_ZERO(&rd_set);

  term = 0;
  clients = 0;
  cli_idx = -1;
  pid = getpid();
  //	for (int i=0; i<CLIENTS_MAX; i++) cpids[i]=0;

  ip = htonl(INADDR_ANY);
  port = 0;
  while (1) {
    char c = getopt_long(argc, argv, "hndai:p:", long_options, NULL);
    if (c == -1)
      break;
    switch (c) {
    case 'h':
      printf("Usage: qscand <options>\n");
      printf("\n");
      printf("-h --help           you are reading this:)\n");
#ifdef DAEMON_EN
      printf("-n --nodaemon       don't daemonize\n");
#endif
      printf("-i --iface <ip>     listen on interface with selected address\n");
      printf("-p --port <port>    listen selected port\n");
#ifdef AUTOPORT_ENABLE
      printf("-a --portauto       autoselect port listen to\n");
#endif
      printf("-d --debug          debug\n");
      printf("\n");
      exit(0);
    case 'n':
      nodaemon = 1;
#ifndef DAEMON_EN
      printf("daemon mode not available, option '-n' ignored!\n");
#endif
      break;
    case 'd':
      debug = 1;
      break;
    case 'i':
      iface = optarg;
      break;
    case 'p': {
      unsigned long lp = strtol(optarg, NULL, 10);
      if (errno == ERANGE || lp > 65535) {
        printf("port number out of range!\n");
      } else {
        port = lp;
      }
    } break;
#ifdef AUTOPORT_ENABLE
    case 'a':
      portauto = 1;
      break;
#endif
    default:
      printf("Invalid option: -%c\n", c);
      break;
    }
  }
  if (!port && !portauto)
    port = DEFPORT;
#ifdef _WIN32
  if ((errno = WSAStartup(MAKEWORD(1, 1), &WSAdata))) {
    printf("Can't initialize winsock: [%d] %s\n", errno, strerror(errno));
    return -1;
  }
#endif
  if ((listenfd = socket(AF_INET, SOCK_STREAM, 0)) == 0) {
    printf("%d: Can't create socket. Terminating\n", pid);
    return -1;
  }
  //	fcntl(listenfd, F_SETFL, fcntl(listenfd, F_GETFL) | O_NONBLOCK);

  // listenfd = socket(AF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
  memset(&srvaddr, 0, sizeof(srvaddr));

#if defined(__unix) || defined(__unix__)
  signal(SIGINT, &sigint_handler);
//	signal(SIGCHLD, &sigchld_handler);
#elif defined(_WIN32)
  SetConsoleCtrlHandler(&sigint_handler, 1);
#endif

  srvaddr.sin_family = AF_INET;
  if (!iface || !inet_aton(iface, &srvaddr.sin_addr)) {
    srvaddr.sin_addr.s_addr = ip;
  }
  srvaddr.sin_port = htons(port);

  int on = 1;
  setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on));

  if (bind(listenfd, (struct sockaddr *)&srvaddr, srvaddr_len)) {
#ifdef _WIN32
    errno = GetLastError();
#endif
    //	if (errno) {
    if (errno != EADDRINUSE || !portauto) {
      printf("%d: Can'n bind() %s:%d: [%d] %s\n", pid,
             inet_ntoa(srvaddr.sin_addr), ntohs(srvaddr.sin_port), errno,
             strerror(errno));
      return -1;
    } else {
      printf("%d: Can'n bind() %s:%d, trying to autoassig port...\n", pid,
             inet_ntoa(srvaddr.sin_addr), ntohs(srvaddr.sin_port));
      srvaddr.sin_port = 0;
      if (bind(listenfd, (struct sockaddr *)&srvaddr, srvaddr_len)) {
        printf("%d: Can'n bind(): %s\n", pid, strerror(errno));
        return -1;
      }
    }
  }

  if (listen(listenfd, LISTENQ)) {
#ifdef _WIN32
    errno = GetLastError();
#endif
    printf("%d: Can'n listen() %s:%d: [%d] %s\n", pid,
           inet_ntoa(srvaddr.sin_addr), ntohs(srvaddr.sin_port), errno,
           strerror(errno));
    return -1;
  }
  //	printf("listenfd: %d\n", listenfd);
  getsockname(listenfd, (struct sockaddr *)&srvaddr, &srvaddr_len);

#ifdef DAEMON_EN
  if (!nodaemon)
    daemonize();
  pid = getpid();

  if (daemonized) {
    syslog(LOG_INFO, "Listening %s:%d\n", inet_ntoa(srvaddr.sin_addr), port);
    syslog(LOG_DEBUG, "daemonized\n");
  } else
#endif
    printf("%d: Listening %s:%d\n", pid, inet_ntoa(srvaddr.sin_addr), port);

  cmutex = new Mutex();
  child_list_clear();

  for (; !term;) {
    //	printf("listen loop...\n");
    socklen_t cliaddr_len = sizeof(childs[cli_idx].arg.cliaddr);
    FD_SET(listenfd, &rd_set);
    tv.tv_sec = 1;
    tv.tv_usec = 0;

    cli_idx = child_find_unused();
    if (cli_idx < 0)
      continue;

    childs[cli_idx].arg.connfd = 0;
    //	printf("select...\n");
    int sret = select(listenfd + 1, &rd_set, NULL, NULL, &tv);
    //	printf("Sret = %d, err: %s\n", sret, strerror(errno));
    if (sret > 0) {
      if (FD_ISSET(listenfd, &rd_set)) {
        // #ifndef _WIN32
        childs[cli_idx].arg.connfd =
            accept(listenfd, (struct sockaddr *)&childs[cli_idx].arg.cliaddr,
                   &cliaddr_len);
        /*
        #else
                                        int tconnfd = accept(listenfd,(struct
        sockaddr*) &childs[cli_idx].arg.cliaddr, &cliaddr_len); #endif
        */
        if (errno == ECONNABORTED) {
          childs[cli_idx].arg.used = 0;
          continue;
        }
#if 0
//#ifdef _WIN32
                //	childs[cli_idx].arg.connfd = _get_osfhandle( tconnfd );
                childs[cli_idx].arg.connfd = _open_osfhandle( tconnfd, _O_RDWR | _O_BINARY | _O_SEQUENTIAL);
                // write(childs[cli_idx].arg.connfd, "012345678\n", 10);
                // write(tconnfd, "012345678\n", 10);
                printf("connfd: %d, errno: [%d] %s\n", childs[cli_idx].arg.connfd, errno, strerror(errno));
#endif
      }
    }
    // if (errno == EINTR) continue;
    if (childs[cli_idx].arg.connfd > 0) {
      //		write(connfd, IDENT, IDENT_LEN);
      //		fcntl(connfd, F_SETFL, fcntl(connfd, F_GETFL) |
      // O_NONBLOCK);
      cmutex->lock();
      if (clients >= CLIENTS_MAX) {
        write(childs[cli_idx].arg.connfd, "QSCAND: clients limit reached!\n",
              31);
        close(childs[cli_idx].arg.connfd);
        childs[cli_idx].arg.used = 0;
        cmutex->unlock();
        continue;
      }
      cmutex->unlock();
      if (!child_create(cli_idx)) {
        cmutex->lock();
        clients++;
        cmutex->unlock();
        if (debug && !daemonized)
          printf("%d: Clients: %d\n", pid, clients);
        /*		for (int i=0; i<CLIENTS_MAX && cpids[i]!=cpid ; )
                        { if (!cpids[i]) cpids[i]=cpid; else i++; } */
      } else {
#ifdef DAEMON_EN
        if (daemonized)
          syslog(LOG_ERR, "Error creating child process\n");
        else
#endif
          printf("%d: Error creating child process\n", pid);
      }
    } else {
      childs[cli_idx].arg.used = 0;
    }
  }

  if (debug && !daemonized)
    printf("%d: Closing listen socket...\n", pid);
  close(listenfd);

  printf("%d: Waiting for childs: %d remains...\n", pid, clients);
  for (int i = 0; i < CLIENTS_MAX; i++) {
    if (!childs[i].arg.used)
      continue;
    //		printf("waiting for thread %d. used: %d\n", i,
    // childs[i].arg.used);
    if (thread_join(childs[i].tid, NULL)) {
#ifdef DAEMON_EN
      if (daemonized)
        syslog(LOG_ERR, "Error waiting for child %d\n", i);
      else
#endif
        printf("Error waiting for child %d\n", i);
    }
  }
#ifdef _WIN32
  WSACleanup();
#endif

  delete cmutex;
#ifdef DAEMON_EN
  if (daemonized)
    syslog(LOG_INFO, "Listener exit\n");
  else
#endif
    printf("%d: Listener exit\n", pid);
#ifdef DAEMON_EN
  if (daemonized)
    closelog();
#endif
  return 0;
}
